apps/argos_aggregation/lib/argos_aggregation/data_sources/thesauri.ex

```diff
@@ -85,44 +242,179 @@ defmodule ArgosAggregation.Thesauri do
-      |> Enum.map(fn pref_label ->
+      defp read_labels(pref_label) do
         %{
           "lang" => xpath(pref_label, ~x(./@xml:lang)s),
           "text" => xpath(pref_label, ~x(./text(\))s)
         }
-      end)
-      |> case do
-        [] ->
-          xml
-          |> xpath(~x(//rdf:Description[@rdf:about="#{@base_url}/#{id}"]/skosxl:literalForm)l)
-          |> Enum.map(fn pref_label ->
-            %{
-              "lang" => xpath(pref_label, ~x(./@xml:lang)s),
-              "text" => xpath(pref_label, ~x(./text(\))s)
-            }
-          end)
+      end
 
-        val ->
-          val
+    end
+
+    defmodule Utils do
+      @base_url Application.get_env(:argos_aggregation, :thesauri_url)
+
+      def check_validity(xml) do
+        try do
+         doc = xml |> parse()
+         {:ok, doc}
+        catch
+          :exit, _ -> {:error, "Malformed xml document"}
+        end
       end
-      |> case do
-        [] ->
-          Logger.warning("No labels found for concept #{@base_url}/#{id}.")
-          []
 
-        val ->
-          val
+      def get_resource_id_from_uri("#{@base_url}/search.rdf" <> _), do: {:error, :search} #wrong url
+      def get_resource_id_from_uri("#{@base_url}/hierarchy/" <> id ), do: {:ok, id}
+      def get_resource_id_from_uri("#{@base_url}/" <> id ), do: {:ok, id}
+      def get_resource_id_from_uri(charlist) when is_list(charlist), do: charlist |> to_string |> get_resource_id_from_uri
+      def get_resource_id_from_uri(error), do: {:error, error}
+    end
+
+    defmodule Search do
+      def load_first_page_url(xml) do
+        case Utils.check_validity(xml) do
+          {:ok, xml} -> xpath(xml, ~x(//sdc:first/@rdf:resource))
+          error -> error
+        end
+      end
+
+      def load_next_page_items(xml) do
+        case Utils.check_validity(xml) do
+         {:ok, xml } ->
+          next_page = xpath(xml, ~x(//sdc:next/@rdf:resource)o)
+          descr_list = xpath(xml, ~x(//rdf:Description[descendant::sdc:link])l)
+          {descr_list, next_page}
+         error -> {[error], nil}
+        end
       end
+
+      def parse_single_doc(doc) do
+        response =
+          doc
+          |> xpath(~x"//sdc:link/@rdf:resource")
+          |> Utils.get_resource_id_from_uri
+        case response do
+          {:ok, id} -> doc |> Factory.assemble_concept(id)
+          error -> error
+        end
+      end
+
     end
+
+    defmodule Hierarchy do
+      def read_root_level(xml) do
+        case Utils.check_validity(xml) do
+          {:ok, xml} ->
+            xml
+            |> xpath(~x"//@rdf:about"l)
+            |> MapSet.new
+            |> MapSet.to_list
+          error -> error
+        end
+      end
+
+      def read_list_of_descriptions(xml) do
+        xml
+        |> xpath(~x"//rdf:Description"l)
+        |> Enum.map(&parse_single_doc(&1))
+      end
+
+      def parse_single_doc(doc) do
```

Comment:
Wieder eine Benennungsfrage: Wieso gibt es in Hierarchy und in Search parse_single_doc? Wann wird was benutzt?

Reply:
Das wird in den Streams benutzt. Wie oben beschrieben tun die Funktionen dasselbe, greifen aber auf eine andere xml Struktur zu und müssen deshalb andere xpath Befehle ausführen.